grub-core/net/drivers/efi/efinet.c

```diff
@@ -373,11 +373,30 @@ grub_efi_net_config_real (grub_efi_handle_t hnd, char **device,
 	if (!match)
 	  continue;
       }
+
     pxe = grub_efi_open_protocol (hnd, &pxe_io_guid,
 				  GRUB_EFI_OPEN_PROTOCOL_GET_PROTOCOL);
     if (! pxe)
       continue;
+
     pxe_mode = pxe->mode;
+
+    struct grub_net_bootp_packet * dhcp_ack    =  &pxe_mode->dhcp_ack;
+    struct grub_net_bootp_packet * proxy_offer =  &pxe_mode->proxy_offer;
+
+    const char server_ip_buff[16];
+
+    grub_ip_uint_to_string(server_ip_buff, dhcp_ack->server_ip);
+    if (grub_strncmp(server_ip_buff, "0.0.0.0", 16) == 0) {
```

Comment:
I see that shim instead checks if a proxy offer was received and the DHCP ack boot file path is not set, and then takes the DHCP ack from the proxy DHCP offer:

https://github.com/rhboot/shim/commit/5f4fd5364109c80934b7837255ddde61f572fd69

I wonder if we shouldn't use the same logic for GRUB.

Reply:
Ahh is that is likely another sane way to check for the condition.

Reply:
```patch
diff --git a/grub-core/net/drivers/efi/efinet.c b/grub-core/net/drivers/efi/efinet.c
index 75e700463..1d8a2c7fc 100644
--- a/grub-core/net/drivers/efi/efinet.c
+++ b/grub-core/net/drivers/efi/efinet.c
@@ -384,12 +384,8 @@ grub_efi_net_config_real (grub_efi_handle_t hnd, char **device,
     struct grub_net_bootp_packet * dhcp_ack    =  &pxe_mode->dhcp_ack;
     struct grub_net_bootp_packet * proxy_offer =  &pxe_mode->proxy_offer;

-    const char server_ip_buff[16];
-
-    grub_ip_uint_to_string(server_ip_buff, dhcp_ack->server_ip);
-    if (grub_strncmp(server_ip_buff, "0.0.0.0", 16) == 0) {
-      // The server IP was 0.0.0.0 which is the same as unset.
-      // This can't be right, we must have gotten here from a proxy DHCP offer
+    if (proxy_offer && dhcp_ack->boot_file[0] == '\0') {
+      // Here we got a proxy offer and the dhcp_ack has a nil boot_file
       // Copy the proxy DHCP offer details intop the bootp_packet we are sending
       // forward as they are the deatils we need.
       *dhcp_ack->server_name = *proxy_offer->server_name;
```

This also seems to work fine in my testing.